### PR TITLE
Group AKS2 files into uncompressed ZIP archive

### PR DESCRIPTION
**Pull request by xinyazhang**
# Overview

To avoid exposing non-ascii characters to file systems. This PR replaces
per-functional `FONLY__*.aks2` files under `aotriton.images/` with **per-kernel
uncompressed ZIP archives** (`<vendor-arch>/<family>/<kernel>.zip`), and add a
zip directory parser `lszip.h` to handle all possible characters in file names.

## Major Changes

* [flatzip] Pack all functionals of a kernel into one installed
  `<vendor-arch>/<family>/<kernel>.zip` (STORED, uncompressed), keyed by
  `unified_signature`. Affine kernels share the layout via a per-family
  `<vendor-arch>/<family>/affine_kernels.zip` so all modules of one family
  land in a single archive.
  + STORED entries are addressable by bare `lseek` + `read` at fixed
    offsets, eliminating zlib dependency. Compression is handled by AKS2
    format.
* [shim] `PackedKernel::open` now handles flatzip through `lszip`
  + Now `PackedKernel` caches both zip directory and hsaco files uncompressed from AKS2.
  + Compressed AKS2 files are not cached. They can be read at any time given
    the directory information.
  + Heterogeneous lookup (e.g. `PStringHash`) is used to support both `string`
    and `string_view`
* [codegen] `--selective` now is `str` rather than `Path` to accepts glob
  patterns (e.g. `flash/affine/*`).
  +  Glob is required for affine kernels,  because all affine kernels share the
     same `affine_kernels.zip`.  Launching multiple workers will overwrite
     others' output.

## Minor Changes

* [codegen] replace `Functional`'s `filepack_signature` / `full_filepack_path`
  properties to distinguish AKS2's on-disk name and in-zip name:
  + `filepack_ondisk_path` → `<vendor-arch>/<family>/<kernel>/<sha256>`
    (intermediate AKS2 stem under `aks2/`)
  + `filepack_inzip_name` → `unified_signature` (ZIP entry name)
  + `full_flatzip_path` → `<vendor-arch>/<family>/<kernel>.zip`
  + `tunecc_signature` adopts `#F;<unified_signature>;;arch=<arch>`
    (was `FONLY__...___<arch>`)
* [codegen] `RootGenerator.write_cluster` becomes path-agnostic (no
  hard-coded `aotriton.images`) so it can be used for
  `Bare.cluster`/`Affine.cluster` and the newly added `Bare.flatzip` shard,
  + Related `.nsv` manifest is also generated accordingly
* [codegen] Replace `package_path` with `flatzip_path` + `aks2_entry`
  across the shim pipeline (`autotune.py`, `autotune_table_entry.cc`,
  `shim.h` context, `shim.cc` `invoke()`); `func_name` adopts
  `unified_signature` to match the new ZIP entry naming.
* [shim] Add `flatzip_path` to `OnDiskKernelInfo`, and replaces `package_path` with `flatzip_path`
  + `OnDeviceKernel::decompress_kernel` now accepts `OnDiskKernelInfo` instead
    of `package_path+stem_name`.
* [shim] Add `lszip.h` / `lszip.cc`: stateless ZIP central-directory parser.
* [build] Move AKS2 files to an intermediate `aks2/` directory;
  `aotriton.images/` now stores the per-kernel `.zip` files
* [doc] Document why `Python3_EXECUTABLE` should be used over `VENV_BIN_PYTHON`
  when finding `pybind11`.
* [doc] Mark `TritonKernel::direct_invoke` with `[[maybe_unused]]`. The
  classical affine kernel is not built anymore

### CODE REVIEW

**Comment by xinyazhang** [comment]
Fast UT passed
```
==> oput_pass0.out <==
[gw6] [ 99%] PASSED test/test_backward.py::test_fast[V3-True-l1-dtype2-0.5-CausalOn-1063-2048-hdim64-(10,2)-3] 
[gw7] [ 99%] PASSED test/test_backward.py::test_fast[V3-True-l1-dtype2-0.5-CausalOn-1063-2048-hdim(120,8)-(10,2)-3] 
[gw4] [ 99%] PASSED test/test_backward.py::test_fast[V3-True-l1-dtype2-0.5-CausalOn-1063-2048-hdim(120,8)-5-3] 
[gw1] [ 99%] PASSED test/test_backward.py::test_fast[V3-True-l1-dtype2-0.5-CausalOn-1063-2048-hdim184-5-3] 
[gw0] [ 99%] PASSED test/test_backward.py::test_fast[V3-True-l1-dtype2-0.5-CausalOn-1063-2048-hdim(24,152)-(10,2)-3] 
[gw2] [ 99%] PASSED test/test_backward.py::test_fast[V3-True-l1-dtype2-0.5-CausalOn-1063-2048-hdim(24,152)-5-3] 
[gw5] [ 99%] PASSED test/test_backward.py::test_fast[V3-True-l1-dtype2-0.5-CausalOn-1063-2048-hdim184-(10,2)-3] 
[gw3] [100%] PASSED test/test_backward.py::test_large_bf16_nan_values[V3-16] 

============================================================================================================================================================================================ 1084 passed in 24.85s =============================================================================================================================================================================================

==> oput_pass0.varlen.out <==
[gw1] [ 99%] PASSED test/test_varlen.py::test_op_bwd[strided-l1-dtype2-0.5-CausalOn-17-hdim8-3] 
[gw0] [ 99%] PASSED test/test_varlen.py::test_op_bwd[strided-l1-dtype2-0.5-CausalOn-17-hdim64-3] 
[gw3] [ 99%] PASSED test/test_varlen.py::test_op_bwd[strided-l1-dtype2-0.5-CausalOn-22-hdim8-3] 
[gw4] [ 99%] PASSED test/test_varlen.py::test_op_bwd[strided-l1-dtype2-0.5-CausalOn-17-hdim(120,8)-3] 
[gw2] [ 99%] PASSED test/test_varlen.py::test_op_bwd[strided-l1-dtype2-0.5-CausalOn-22-hdim184-3] 
[gw6] [ 99%] PASSED test/test_varlen.py::test_op_bwd[strided-l1-dtype2-0.5-CausalOn-22-hdim64-3] 
[gw5] [ 99%] PASSED test/test_varlen.py::test_op_bwd[strided-l1-dtype2-0.5-CausalOn-22-hdim(24,152)-3] 
[gw7] [100%] PASSED test/test_varlen.py::test_op_bwd[strided-l1-dtype2-0.5-CausalOn-22-hdim(120,8)-3] 

============================================================================================================================================================================================= 900 passed in 22.01s =============================================================================================================================================================================================
```

File: include/aotriton/_internal/on_device_kernel.h [hunk 1]
```diff
@@ -32,9 +32,10 @@ class OnDeviceKernel {
...
   };
   struct OnDiskKernelInfo {
-    pstring_view package_path;
-    std::string_view stem_name;
-    std::string_view function_name;
+    pstring_view     flatzip_path;  // path to .zip relative to aotriton.images/
+    std::string_view aks2_entry;    // ZIP entry name = unified_signature
+    std::string_view stem_name;     // HSACO entry name inside AKS2 (;;#F/P/CO/arch)
+    std::string_view function_name; // HIP kernel symbol name
   };
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply include/aotriton/_internal/on_device_kernel.h:39]: `direct_invoke` is not used ATM

File: v3src/CMakeLists.txt [hunk 2]
```diff
@@ -216,65 +216,120 @@ if(NOT AOTRITON_NOIMAGE_MODE)
...
+        ${DIR_ARCH}
+        ${DIR_FAMILY}
+        ${DIR_KERNEL}
+        "${SHA256}.nsv")
+      # message(STATUS "Add AKS2 ${AKS2}")
+      add_custom_command(OUTPUT "${AKS2}"
+        COMMAND ${CMAKE_COMMAND} -E env VIRTUAL_ENV=${VENV_DIR}
+        "${VENV_BIN_PYTHON}"
+        -m v3python.aks2
+        -o "${AKS2}"
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply v3src/CMakeLists.txt:250]: The directory is created by ninja automatically

File: v3src/packed_kernel_common.cc [hunk 3]
```diff
@@ -172,14 +203,19 @@ PackedKernel::PackedKernel(int fd) {
   strm.next_out = (uint8_t*)decompressed_content_.data();
   strm.avail_out = decompressed_content_.size();
   lzma_action action = LZMA_RUN;
+  // Track remaining bytes when size is bounded (reading AKS2 from inside a ZIP).
+  size_t remaining = (size == SIZE_MAX) ? SIZE_MAX : (size - sizeof(AKS2_Header));
   while (true) {
     if (strm.avail_in == 0) {
       strm.next_in = inbuf;
```
[the remark this answers is not in the document]
Comment (xinyazhang) [reply v3src/packed_kernel_common.cc:223]: XZ format is self-delimiting and don't require LZMA_FINISH

**Review by jeffdaily — CHANGES_REQUESTED** [review]
## Review: Group AKS2 files into uncompressed ZIP archive

Nice design — STORED-so-`lseek`-addressable is a clean way to drop the zlib dependency at read time. A few blockers in the new ZIP parser and missing test coverage.

### Security

**`v3src/lszip.cc` — heap OOB read on entry name (must fix).** The central-directory loop validates only the fixed 46-byte record (`if (p + 46 > end ...)`), then constructs `entry_name(reinterpret_cast<const char*>(p + 46), fname_len)` (and the verbose-log `std::string_view` just above it) without checking `p + 46 + fname_len <= end`. `fname_len`, `extra_len`, `comment_len`, and `total_entries` all come from the file, so a truncated/corrupted central directory yields a read of up to ~64 KB past the `cd` buffer — handed to the visitor and copied into a `std::string`. Suggest adding, before reading the name:

```cpp
if (p + 46 + fname_len + extra_len + comment_len > end)
  break;
```

This is the same corrupted-input class the EOCD `cd_offset + cd_size <= file_size` guard already defends against; the per-entry case was missed.

### Code Quality

**`v3src/lszip.cc` — `break` on an unexpected entry silently truncates the directory.** On a compressed/encrypted entry (`comp_method != 0 || gp_flag & 1`) or a short local-header read, the loop `break`s. But `PackedKernel::open`'s contract is "InnerMap is fully populated from the ZIP central directory on first open, so `inner == end` is ground truth that the entry is absent." A `break` violates that invariant: every valid entry *after* the offending one is dropped from the map and later reported as a misleading not-found/`nullptr` at runtime rather than a parse error. STORED-only construction makes this defensive today, but if it ever fires it fails silently — prefer a hard failure signal so a partial map is never cached as authoritative.

**`v3src/packed_kernel_common.cc` — redundant registry lookups.** The slow path does `registry_.contains(flatzip_path)`, then `registry_[pstring_type(flatzip_path)]` inside the populate block, then `registry_[pstring_type(flatzip_path)]` again for `inner_map` — up to three hashes plus two full-string `pstring_type` copies for one logical access. Hold a single `InnerMap&`.

### Testing

**No tests for the new parser or packer (must fix).** `lszip` and `v3python/flatzip.py` are self-contained and unit-testable, and the bounds/robustness edge cases above are exactly what a parser needs covered. Minimum: a `flatzip.py` → `lszip` round-trip asserting `(name, offset, size)` per entry, plus malformed-input cases (truncated central directory, oversized `fname_len`, non-STORED entry) that must be rejected without OOB access.

### Infrastructure

**`v3src/CMakeLists.txt` — ZIP staleness may not track its real inputs.** The per-kernel `.zip` `add_custom_command` lists `DEPENDS aotriton_v3_aks2 "${ZIP_MANIFEST}"`, but the actual inputs are the individual `.aks2` files named *inside* the `.nsv` manifest. Depending on the aggregate target plus an unchanged manifest can leave a stale `.zip` when a single `.aks2` is regenerated without the manifest text changing. Please confirm a single-kernel content change triggers a `.zip` rebuild, or depend on the concrete `.aks2` paths.

### Minor

- `v3python/flatzip.py` — offsets/sizes/CRC are packed as 32-bit `'I'` with no ZIP64. A cumulative offset >4 GiB raises `struct.error` at build time (loud, not silent corruption), so it's acceptable today; worth a comment noting the STORED/no-ZIP64 ceiling now that archives aggregate all functionals of a kernel.
- `v3python/flatzip.py` — `_VERSION_NEEDED = 20` is labeled "minimum for deflate" but the archive is STORED (only needs 1.0 / version 10). Cosmetic.

---

Blocking on the `lszip` heap OOB read and the missing parser/packer tests. The `break`-truncation invariant and the CMake staleness question should be resolved or explicitly dismissed before merge.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by xinyazhang** [comment]
Fix. We don't need separate tests for packer and parser. The success of current UT implies the correctness of packer+parser.

**Review by jeffdaily — APPROVED** [review]
